Mathlib/Algebra/CharP/Lemmas.lean

```diff
@@ -25,29 +25,36 @@ include hp
 protected theorem add_pow_prime_pow_eq (h : Commute x y) (n : ℕ) :
     (x + y) ^ p ^ n =
       x ^ p ^ n + y ^ p ^ n +
-        p * ∑ k ∈ Ioo 0 (p ^ n), x ^ k * y ^ (p ^ n - k) * ↑((p ^ n).choose k / p) := by
-  trans x ^ p ^ n + y ^ p ^ n + ∑ k ∈ Ioo 0 (p ^ n), x ^ k * y ^ (p ^ n - k) * (p ^ n).choose k
-  · simp_rw [h.add_pow, ← Nat.Ico_zero_eq_range, Ico_add_one_right_eq_Icc,
-      Icc_eq_cons_Ico (zero_le _), Finset.sum_cons, Ico_eq_cons_Ioo (pow_pos hp.pos _),
-      Finset.sum_cons, tsub_self, tsub_zero, pow_zero, Nat.choose_zero_right, Nat.choose_self,
-      Nat.cast_one, mul_one, one_mul, ← add_assoc]
-  · congr 1
-    simp_rw [Finset.mul_sum, Nat.cast_comm, mul_assoc _ _ (p : R), ← Nat.cast_mul]
-    refine Finset.sum_congr rfl fun i hi => ?_
-    rw [mem_Ioo] at hi
-    rw [Nat.div_mul_cancel (hp.dvd_choose_pow hi.1.ne' hi.2.ne)]
+        p * x * y *
+          ∑ k ∈ Ioo 0 (p ^ n), x ^ (k - 1) * y ^ (p ^ n - k - 1) * ↑((p ^ n).choose k / p) := calc
+  _ = ∑ k ∈ Icc 0 (p ^ n), x ^ k * y ^ (p ^ n - k) * (p ^ n).choose k := by
+    rw [h.add_pow, ← Nat.Ico_zero_eq_range, Ico_add_one_right_eq_Icc]
+  _ = x ^ p ^ n + y ^ p ^ n + ∑ k ∈ Ioo 0 (p ^ n), x ^ k * y ^ (p ^ n - k) * (p ^ n).choose k := by
+    simp_rw [Icc_eq_cons_Ico (zero_le _), Ico_eq_cons_Ioo (pow_pos hp.pos _)]
+    simp [-cons_eq_insert, add_assoc]
+  _ = _ := by
+    simp_rw [Finset.mul_sum]
+    congr! 2 with k hk
+    obtain ⟨hk₀, hk⟩ := mem_Ioo.1 hk
+    rw [← mul_pow_sub_one (by omega), ← mul_pow_sub_one (n := p ^ n - k) (by omega)]
+    -- The maths is over now. We just commute things to their place.
+    rw [Nat.cast_comm, mul_assoc x _ y, Nat.cast_comm, ← mul_assoc x y,
+      (p.commute_cast _).symm.mul_mul_mul_comm (x * y), (h.pow_left _).symm.mul_mul_mul_comm]
+    norm_cast
+    rw [Nat.mul_div_cancel' (hp.dvd_choose_pow _ _)] <;> omega
 
 protected theorem add_pow_prime_eq (h : Commute x y) :
     (x + y) ^ p =
-      x ^ p + y ^ p + p * ∑ k ∈ Finset.Ioo 0 p, x ^ k * y ^ (p - k) * ↑(p.choose k / p) := by
+      x ^ p + y ^ p + p * x * y *
+        ∑ k ∈ Finset.Ioo 0 p, x ^ (k - 1) * y ^ (p - k - 1) * ↑(p.choose k / p) := by
```

Comment:
Ditto here. One way of summarising what you've done is that modulo a trivial lemma about commuting multiplication and a sum, you're replacing x^k with x^{1+k-1} in a situation where one can check that all the summands have k>0 (and perhaps even more annoyingly replacing y^{p-k} with y^{1+(p-k-1)}). But x^k looks more natural to me.

I am not objecting to the new statements, I think they're useful, and clearly it's unquestionable that `exists_add_pow_prime_pow_eq` has got better because the new claim is strictly stronger. It's just not clear to me that deleting what might be just the right lemma for someone and replacing it by something which might not be more convenient for everyone is what we should be doing here.

Can you just put the old ones back? e.g. deduce them from the new ones?

Reply:
> One way of summarising what you've done is that modulo a trivial lemma about commuting multiplication and a sum, you're replacing x^k with x^{1+k-1} in a situation where one can check that all the summands have k>0 (and perhaps even more annoyingly replacing y^{p-k} with y^{1+(p-k-1)}). But x^k looks more natural to me.

I would usually agree, but this statement is clearly meant to provide the witness of `exists_add_pow_prime_pow_eq` explicitly: You are complaining that I am introducing truncated subtraction in the statement, but there is already truncated division in `p.choose k / p`!

Reply:
I do not believe the point about truncated division is relevant.

However I am convinced by the argument about the job of this lemma being to provide a witness for `exists_add_pow_prime_pow_eq`. Furthermore I think that if we add `add_pow_prime_pow_eq'` above then the old statement is very, very easy to recover in the event that someone ever wants it.